app/public/cantusdata/helpers/source_importer.py

```diff
@@ -0,0 +1,124 @@
+from urllib import request
+import json
+from html.parser import HTMLParser
+
+class ProvenanceParser(HTMLParser):
+    def __init__(self):
+        super().__init__()
+        self.provenance_class = False
+        self.is_provenance = False
+        self.provenance = ""
+
+    def handle_starttag(self, tag, attrs):
+        if tag == "div":
+            if ("class","views-field views-field-field-provenance-tax") in attrs:
+                self.provenance_class = True
+        if tag == "a":
+            if self.provenance_class:
+                self.is_provenance = True
+
+    def handle_endtag(self, tag):
+        if self.is_provenance:
+            self.provenance_class = False
+            self.is_provenance = False
```

Comment:
I don't really understand what's happening here. What is `self.provenance_class` meant to indicate?

Reply:
`self.provenance_class` is a flag that is set to `True` when the parser reaches the element in the html document that contains the provenance data that we need to extract. 

The relevant section of the source page html looks like this:  
```
<div class="views-field views-field-field-provenance-tax">    
    <span class="views-label views-label-field-provenance-tax">Provenance: </span>    
    <strong class="field-content">
        <a href="/provenance/3666" typeof="skos:Concept" property="rdfs:label skos:prefLabel" datatype="">Eichstätt</a>
    </strong>  
</div>
```

Reply:
> Do note that you'll be looking for `/provenance/id-of-provenance/` rather than `.../id-of-source/`.

Good point!